Show evaluated expr when `@test isequal(...)` fails

omus opened the pull request:
Support showing the evaluated version of the expression when using `@test` with `isequal`:

```julia
julia> using Base.Test

julia> x, y = 1, 2
(1, 2)

julia> @test isequal(x, y)
Test Failed
  Expression: isequal(x, y)
   Evaluated: isequal(1, 2)
ERROR: There was an error during testing
```

omus committed:
Show evaluated expr with `@test isequal(...)`

tkelman commented on base/test.jl:
```diff
@@ -305,6 +305,12 @@ function get_test_result(ex)
             terms[i] = esc(terms[i])
         end
         testret = :(eval_comparison(Expr(:comparison, $(terms...))))
+    elseif isa(ex, Expr) && ex.head == :call && ex.args[1] == :isequal
```
`ex.args[1] in (:isequal, :isapprox, :≈)` would also be useful if that comes out looking right

omus replied:
The `≈` operator is evaluated already. Unfortunately it isn't quite right:

```julia
julia> @test x ≈ y
Test Failed
  Expression: x ≈ y
   Evaluated: 1 isapprox 2
```

omus replied:
Seems like this wasn't tested very thoroughly:
```julia
julia> @test x ≈ y ≈ z
Test Failed
  Expression: $(Expr(:escape, :x)) $(Expr(:escape, :≈)) $(Expr(:escape, :y)) $(Expr(:escape, :≈)) $(Expr(:escape, :z))
   Evaluated: 1 isapprox 2 isapprox 3
```
I'll try to address this.

omus replied:
Addressed all of the above mentioned issues.

omus committed:
Show evaluated expr with `@test isapprox(...)`

omus commented:
Changes now include:
* Corrected `@test x == y == z` shown expression
* Corrected `@test x ≈ y` shown evaluated expression
* Added showing evaluated expression of `@test isequal(...)` and `@test isapprox(...)`
* Lots of tests

KristofferC commented:
Any commits that can be backported?

omus commented:
> Any commits that can be backported?

I'll break this into two PRs which separates the new features from the bugfixes

omus commented:
I couldn't make a clean separation between the fixes and the features. PR #22315 contains just the fixes and this PR extends that work with the features.

omus commented:
Sibling PR #22315 has been merged. I've rebased this PR which now makes the diff understandable. Reviews are welcome.

omus commented:
I think this is good to go. I'll merge at the end of JuliaCon unless there are objections.

nalimilan commented:
This seems to have broken StatsBase tests. MWE:
```julia

julia> @test isapprox(1, 1, atol=0)
Error During Test
  Test threw an exception of type MethodError
  Expression: isapprox(1, 1, atol=0)
  MethodError: no method matching Expr(::Symbol, ::Base.#isapprox, ::Int64, ::Int64; atol=0)
  Closest candidates are:
    Expr(::ANY...) at boot.jl:193 got unsupported keyword argument "atol"
  Stacktrace:
   [1] eval(::Module, ::Any) at ./boot.jl:236
   [2] eval_user_input(::Any, ::Base.REPL.REPLBackend) at ./repl/REPL.jl:67
   [3] macro expansion at ./repl/REPL.jl:98 [inlined]
   [4] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:73
ERROR: There was an error during testing
```

omus commented:
I'll look into it.

omus commented:
I just noticed that I need to some additional work to get splatting to work with the call syntax:
```julia
julia> @test isequal((1,1)...)
Error During Test
  Test threw an exception of type BoundsError
  Expression: isequal((1, 1)...)
  BoundsError: attempt to access 3-element Array{Any,1} at index [3:4]
```
Additionally, there is a corner case that was introduced in Julia 0.5 with `==`:

```julia
julia> @test ==(1, (1, 2)...)
Error During Test
  Test threw an exception of type BoundsError
  Expression: ==(1,(1,2)...)
  BoundsError: attempt to access 4-element Array{Any,1} at index [5]

julia> ==(1, (1, 2)...)
ERROR: MethodError: no method matching ==(::Int64, ::Int64, ::Int64)
```

The `@test` is failing here as we should see the `MethodError` and not the `BoundsError`.
